File: scripts/export-literary-samples.ts
/**
 * 导出本地小说/漫画样例（精选公开作品 + 关联资源）到目录，供 import-literary-samples 导入生产库。
 *
 *   DATABASE_URL="file:./prisma/dev.db" npx tsx scripts/export-literary-samples.ts
 *   npx tsx scripts/export-literary-samples.ts --out ./data/literary-samples-bundle
 */
import fs from "node:fs/promises";
import path from "node:path";
import { prisma } from "@/lib/prisma";

export const LITERARY_SAMPLES_OWNER = "__literary-samples__";

const SHOWCASE_TITLE_PATTERNS = [
  "煤山",
  "崇祯",
  "锦衣卫",
  "Jinyiwei",
  "錦衣衛",
  "洪荒",
  "废柴世子",
  "孤剑",
  "千亿继承人",
  "聪明的小老鼠",
  "糯米团",
  "花花草草",
  "愚公",
  "时空",
  "第一章迷雾",
  "核爆",
];

function isShowcaseTitle(title: string): boolean {
  return SHOWCASE_TITLE_PATTERNS.some((p) => title.includes(p));
}

function collectPublicPaths(coverPath: string | null, imageUrlsJson: string | null): string[] {
  const out = new Set<string>();
  if (coverPath?.startsWith("/")) out.add(coverPath.slice(1));
  if (imageUrlsJson) {
    try {
      const urls = JSON.parse(imageUrlsJson) as unknown;
      if (Array.isArray(urls)) {
        for (const u of urls) {
          if (typeof u === "string" && u.startsWith("/")) out.add(u.slice(1));
        }
      }
    } catch {
      /* ignore */
    }
  }
  return [...out];
}

async function main(): Promise<void> {
  const outArg = process.argv.indexOf("--out");
  const outDir = path.resolve(
    outArg >= 0 && process.argv[outArg + 1] ? process.argv[outArg + 1] : "data/literary-samples-bundle",
  );
  const repoRoot = process.cwd();

  const novels = await prisma.novel.findMany({ where: { visibility: "public" } });
  const comics = await prisma.comic.findMany({ where: { visibility: "public" } });

  const novelIds = new Set<string>();
  for (const n of novels) {
    if (n.featured || isShowcaseTitle(n.title)) novelIds.add(n.id);
  }
  const comicIds = new Set<string>();
  for (const c of comics) {
    if (c.featured) {
      comicIds.add(c.id);
      if (c.novelId) novelIds.add(c.novelId);
    }
  }
  for (const c of comics) {
    if (c.novelId && novelIds.has(c.novelId)) comicIds.add(c.id);
  }

  const selNovels = novels.filter((n) => novelIds.has(n.id));
  const selComics = comics.filter((c) => comicIds.has(c.id));

  const files = new Set<string>();
  for (const row of [...selNovels, ...selComics]) {
    for (const f of collectPublicPaths(row.coverPath, "imageUrls" in row ? row.imageUrls : null)) {
      files.add(f);
    }
  }

  await fs.rm(outDir, { recursive: true, force: true });
  await fs.mkdir(path.join(outDir, "public"), { recursive: true });

  const copied: string[] = [];
  for (const rel of files) {
    const src = path.join(repoRoot, "public", rel);
    const dest = path.join(outDir, "public", rel);
    try {
      await fs.mkdir(path.dirname(dest), { recursive: true });
      await fs.copyFile(src, dest);
      copied.push(rel);
    } catch {
      console.warn("[warn] missing asset:", rel);
    }
  }

  const manifest = {
    version: 1,
    ownerKey: LITERARY_SAMPLES_OWNER,
    exportedAt: new Date().toISOString(),
    novels: selNovels.map(({ ownerKey: _o, ...rest }) => ({ ...rest, ownerKey: LITERARY_SAMPLES_OWNER })),
    comics: selComics.map(({ ownerKey: _o, ...rest }) => ({ ...rest, ownerKey: LITERARY_SAMPLES_OWNER })),
    files: copied,
  };

  await fs.writeFile(path.join(outDir, "manifest.json"), JSON.stringify(manifest, null, 2), "utf8");

  console.log(
    `[export-literary-samples] novels=${selNovels.length} comics=${selComics.length} files=${copied.length} → ${outDir}`,
  );
}

main()
  .catch((e) => {
    console.error("[FAIL]", e instanceof Error ? e.message : e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

File: scripts/export-literary-samples.py
#!/usr/bin/env python3
"""Export curated novel/comic samples from local dev.db → data/literary-samples-bundle/"""
import json, os, shutil, sqlite3
from datetime import datetime
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DB = ROOT / "prisma" / "dev.db"
OUT = ROOT / "data" / "literary-samples-bundle"
OWNER = "__literary-samples__"

PATTERNS = [
    "煤山", "崇祯", "锦衣卫", "Jinyiwei", "錦衣衛", "洪荒", "废柴世子", "孤剑", "千亿继承人",
    "聪明的小老鼠", "糯米团", "花花草草", "愚公", "时空", "第一章迷雾", "核爆",
]

def showcase(title: str) -> bool:
    return any(p in title for p in PATTERNS)

def row_dict(row, cols):
    return {c: row[i] for i, c in enumerate(cols)}

def collect_paths(cover, image_urls):
    paths = set()
    if cover and str(cover).startswith("/"):
        paths.add(str(cover).lstrip("/"))
    if image_urls:
        try:
            for u in json.loads(image_urls):
                if isinstance(u, str) and u.startswith("/"):
                    paths.add(u.lstrip("/"))
        except Exception:
            pass
    return paths

def main():
    c = sqlite3.connect(DB)
    ncols = [d[1] for d in c.execute("PRAGMA table_info(Novel)")]
    ccols = [d[1] for d in c.execute("PRAGMA table_info(Comic)")]
    novels = [row_dict(r, ncols) for r in c.execute("SELECT * FROM Novel WHERE visibility='public'")]
    comics = [row_dict(r, ccols) for r in c.execute("SELECT * FROM Comic WHERE visibility='public'")]

    novel_ids = {n["id"] for n in novels if n.get("featured") or showcase(n["title"])}
    comic_ids = set()
    for cm in comics:
        if cm.get("featured"):
            comic_ids.add(cm["id"])
            if cm.get("novelId"):
                novel_ids.add(cm["novelId"])
    for cm in comics:
        if cm.get("novelId") in novel_ids:
            comic_ids.add(cm["id"])

    sel_novels = [n for n in novels if n["id"] in novel_ids]
    sel_comics = [cm for cm in comics if cm["id"] in comic_ids]

    files = set()
    for row in sel_novels + sel_comics:
        files |= collect_paths(row.get("coverPath"), row.get("imageUrls"))

    if OUT.exists():
        shutil.rmtree(OUT)
    (OUT / "public").mkdir(parents=True)

    copied = []
    for rel in sorted(files):
        src = ROOT / "public" / rel.replace("/", os.sep)
        dest = OUT / "public" / rel.replace("/", os.sep)
        if not src.is_file():
            print("[warn] missing", rel)
            continue
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dest)
        copied.append(rel)

    for row in sel_novels + sel_comics:
        row["ownerKey"] = OWNER

    manifest = {
        "version": 1,
        "ownerKey": OWNER,
        "exportedAt": datetime.utcnow().isoformat() + "Z",
        "novels": sel_novels,
        "comics": sel_comics,
        "files": copied,
    }
    (OUT / "manifest.json").write_text(json.dumps(manifest, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"[ok] novels={len(sel_novels)} comics={len(sel_comics)} files={len(copied)} -> {OUT}")

if __name__ == "__main__":
    main()
